src/Data/HashTable/Class.hs

```diff
@@ -81,6 +81,15 @@ class HashTable h where
     -- table. /O(n)/.
     mapM_    :: ((k,v) -> ST s b) -> h s k v -> ST s ()
 
+    -- | Looks up the index of a key-value mapping in a hash table suitable
+    -- for passing to 'nextByIndex'.
+    lookupIndex :: (Eq k, Hashable k) => h s k v -> k -> ST s (Maybe Word)
```

Comment:
The docs ought to make very clear that the index returned by `lookupIndex` is only valid until the next `insert` since that can grow the table and move everything about.